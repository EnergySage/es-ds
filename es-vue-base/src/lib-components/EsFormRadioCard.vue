<template>
    <b-form-radio
        class="EsFormRadioCard bg-white border-light rounded-lg w-100"
        :value="value">
        <slot />
    </b-form-radio>
</template>

<script>
import { BFormRadio } from 'bootstrap-vue';

export default {
    name: 'EsFormRadioCard',
    components: {
        BFormRadio,
    },
    props: {
        value: {
            type: String,
            required: true,
            default: '',
        },
    },
};
</script>

<style lang="scss" scoped>
@use "sass:map";
@import '~@energysage/es-bs-base/scss/includes';

.EsFormRadioCard {
    /**
     * overridable styles spelled out in CSS rather than utility classes so users are able to pass in
     * utility classes and have them override the default styles (otherwise, if the user's utility classes
     * happen to appear earlier in the es-bs-base CSS than the component's utility classes, the component's
     * will win, leaving the user unable to override the desired style)
     */
    @include font-size($h2-font-size);
    font-weight: $font-weight-bolder;
    margin-bottom: map.get($spacers, 3);
    padding-bottom: map.get($spacers, 4);
    padding-left: map.get($spacers, 3);
    padding-right: map.get($spacers, 3);
    padding-top: map.get($spacers, 4);
    text-align: left;

    ::v-deep input {
        /**
         * hide the radio button input elements the same way bootstrap vue does.
         * bootstrap vue's styles only work on items directly nested under b-form-radio-group, and we want
         * to enable users to put any elements (e.g. <b-col> and <b-row>) between the EsFormRadioCards and
         * EsFormRadioCard elements.
         */
        clip: rect(0, 0, 0, 0);
        pointer-events: none;
        position: absolute;
    }

    &.btn {
        /* override the fixed height of normal buttons */
        height: auto;
    }

    &:hover {
        /* needed to override the default background/text color switch for outline button hover state */
        background-color: $white;
        color: $dark;
    }

    &:not(:disabled):not(.disabled):active,
    &:not(:disabled):not(.disabled).active,
    .show > &.dropdown-toggle {
        background-color: $white;
        /* override 'border-light' utility class */
        border-color: $dark !important;
        color: $dark;
    }
}

@include media-breakpoint-up(lg) {
    .EsFormRadioCard {
        /**
        * overridable styles spelled out in CSS rather than utility classes so users are able to pass in
        * utility classes and have them override the default styles (otherwise, if the user's utility classes
        * happen to appear earlier in the es-bs-base CSS than the component's utility classes, the component's
        * will win, leaving the user unable to override the desired style)
        */
        @include font-size($h2-font-size-desktop);
        padding-left: map.get($spacers, 4);
        padding-right: map.get($spacers, 4);
    }
}
</style>
