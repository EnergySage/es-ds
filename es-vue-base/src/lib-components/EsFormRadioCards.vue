<template>
    <b-form-group
        v-slot="{ ariaDescribedby }"
        :label="label"
        :label-class="labelClass"
        :label-sr-only="labelSrOnly">
        <b-form-radio-group
            :aria-describedby="ariaDescribedby"
            buttons
            button-variant="outline-primary"
            :checked="value"
            class="es-form-radio-cards d-flex justify-content-center"
            :class="{ 'has-icon': hasIcon }"
            :name="label"
            stacked
            v-bind="$attrs"
            v-on="$listeners">
            <slot />
        </b-form-radio-group>
    </b-form-group>
</template>

<script>
import { BFormGroup, BFormRadioGroup } from 'bootstrap-vue';

export default {
    name: 'EsFormRadioCards',
    components: {
        BFormGroup,
        BFormRadioGroup,
    },
    props: {
        hasIcon: {
            type: Boolean,
            default: false,
        },
        label: {
            type: String,
            required: true,
        },
        labelClass: {
            type: String,
            default: 'font-size-h1 font-weight-bolder mb-200 pb-0 text-dark',
        },
        labelSrOnly: {
            type: Boolean,
            default: false,
        },
        value: {
            type: String,
            required: false,
            default: '',
        },
    },
};
</script>
