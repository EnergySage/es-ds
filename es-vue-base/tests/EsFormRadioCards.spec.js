import { mount } from '@vue/test-utils';
import EsFormRadioCards from '@/src/lib-components/EsFormRadioCards.vue';
import EsFormRadioCard from '@/src/lib-components/EsFormRadioCard.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsFormRadioCards', () => {
    // Basic test; should exist for most components
    test('<EsFormRadioCards />', async () => {
        const wrapper = mount(EsFormRadioCards, {
            ...jestVue,
            propsData: {
                label: 'What is one plus one?',
            },
            slots: {
                default: `
                  <es-form-radio-card value="2">
                      Two
                  </es-form-radio-card>
              `,
            },
            stubs: {
                'es-form-radio-card': EsFormRadioCard,
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    test('EsFormRadioCards card click updates value', async () => {
        const label = 'What is one plus one?';
        const wrapper = mount({
            ...jestVue,
            data() {
                return {
                    answer: '',
                };
            },
            template: `
                <EsFormRadioCards label="${label}" v-model="answer">
                    <EsFormRadioCard value="1">
                        One
                    </EsFormRadioCard>
                    <EsFormRadioCard data-testid="radioCardTwo" value="2">
                        Two
                    </EsFormRadioCard>
                    <EsFormRadioCard value="3">
                        Three
                    </EsFormRadioCard>
                </EsFormRadioCards>
            `,
            components: {
                EsFormRadioCards,
                EsFormRadioCard,
            },
        });

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();

        // expect the data to be the default value
        expect(wrapper.vm.answer).toBe('');

        // "click" one of the radio cards by calling the "change" event on it
        const testRadioCard = wrapper.find('[data-testid="radioCardTwo"]');
        const testRadioButton = testRadioCard.find('input[type="radio"]');
        await testRadioButton.trigger('change');

        // expect that the value corresponding to the "clicked" card is set
        expect(wrapper.vm.answer).toBe('2');
    });
});
