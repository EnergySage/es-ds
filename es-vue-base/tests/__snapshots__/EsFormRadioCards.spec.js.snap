// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsFormRadioCards <EsFormRadioCards /> 1`] = `
"<fieldset class="form-group" id="__BVID__4">
  <legend tabindex="-1" class="bv-no-focus-ring col-form-label pt-0 font-size-h1 font-weight-bolder mb-4 pb-0 text-dark" id="__BVID__4__BV_label_">What is one plus one?</legend>
  <div>
    <div role="radiogroup" tabindex="-1" class="EsFormRadioCards d-flex justify-content-center btn-group-toggle btn-group-vertical bv-no-focus-ring" id="__BVID__5"><label class="EsFormRadioCard bg-white border-light rounded-lg w-100 btn btn-outline-dark"><input type="radio" name="What is one plus one?" class="" value="2" id="__BVID__7">
        Two
      </label></div>
    <!---->
    <!---->
    <!---->
  </div>
</fieldset>"
`;

exports[`EsFormRadioCards EsFormRadioCards card click updates value 1`] = `
"<fieldset class="form-group">
  <legend tabindex="-1" class="bv-no-focus-ring col-form-label pt-0 font-size-h1 font-weight-bolder mb-4 pb-0 text-dark">What is one plus one?</legend>
  <div>
    <div role="radiogroup" tabindex="-1" class="EsFormRadioCards d-flex justify-content-center btn-group-toggle btn-group-vertical bv-no-focus-ring"><label class="EsFormRadioCard bg-white border-light rounded-lg w-100 btn btn-outline-dark"><input type="radio" name="What is one plus one?" class="" value="1">
        One
      </label> <label class="EsFormRadioCard bg-white border-light rounded-lg w-100 btn btn-outline-dark" data-testid="radioCardTwo"><input type="radio" name="What is one plus one?" class="" value="2">
        Two
      </label> <label class="EsFormRadioCard bg-white border-light rounded-lg w-100 btn btn-outline-dark"><input type="radio" name="What is one plus one?" class="" value="3">
        Three
      </label></div>
    <!---->
    <!---->
    <!---->
  </div>
</fieldset>"
`;
